src/CoreWCF.Primitives/src/CoreWCF/Security/DataProtectionSecurityStateEncoder.cs

```diff
@@ -3,64 +3,26 @@
 
 using System;
 using System.Security.Cryptography;
-using System.Text;
-using CoreWCF.Runtime;
+using Microsoft.AspNetCore.DataProtection;
 
 namespace CoreWCF.Security
 {
     public class DataProtectionSecurityStateEncoder : SecurityStateEncoder
     {
-        private readonly byte[] _entropy;
+        private readonly Lazy<IDataProtector> _dataProtector;
 
-        public DataProtectionSecurityStateEncoder() : this(true)
-        {
-            // empty
-        }
-
-        public DataProtectionSecurityStateEncoder(bool useCurrentUserProtectionScope) : this(useCurrentUserProtectionScope, null)
-        { }
-
-        public DataProtectionSecurityStateEncoder(bool useCurrentUserProtectionScope, byte[] entropy)
-        {
-            UseCurrentUserProtectionScope = useCurrentUserProtectionScope;
-            if (entropy == null)
-            {
-                _entropy = null;
-            }
-            else
-            {
-                _entropy = Fx.AllocateByteArray(entropy.Length);
-                Buffer.BlockCopy(entropy, 0, _entropy, 0, entropy.Length);
-            }
-        }
-
-        public bool UseCurrentUserProtectionScope { get; }
+        public DataProtectionSecurityStateEncoder() =>  _dataProtector = new Lazy<IDataProtector>(GetDataProtectorFromDI);
 
-        public byte[] GetEntropy()
-        {
-            byte[] result = null;
-            if (_entropy != null)
-            {
-                result = Fx.AllocateByteArray(_entropy.Length);
-                Buffer.BlockCopy(_entropy, 0, result, 0, _entropy.Length);
-            }
-            return result;
-        }
-
-        public override string ToString()
-        {
-            StringBuilder result = new StringBuilder();
-            result.Append(GetType().ToString());
-            result.AppendFormat("{0}  UseCurrentUserProtectionScope={1}", Environment.NewLine, UseCurrentUserProtectionScope);
-            result.AppendFormat("{0}  Entropy Length={1}", Environment.NewLine, (_entropy == null) ? 0 : _entropy.Length);
-            return result.ToString();
-        }
+        private static IDataProtector GetDataProtectorFromDI()
+            => OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>().GetDataProtector(nameof(DataProtectionSecurityStateEncoder));
+         
```

Comment:
This won't work unless the service instance has been added to DI. The `Find<IServiceProvider>()` will throw null in that case.

Reply:
While it's really unlikely, we need to avoid the possibility of accidental naming collisions. If you use the fully qualified name of this class, the only way to get a collision is if someone else explicitly chooses "CoreWCF.Security.DataProtectionSecurityStateEncoder", which would have to be on purpose.

Reply:
> This won't work unless the service instance has been added to DI. The `Find<IServiceProvider>()` will throw null in that case.

Indeed. I'm wondering what is the best way to get a `IServiceProvider` here ? Should i try to pass it from call sites ? Despite a kind of ServiceLocator would i think do the job here I'm wondering if you would be open to it due to potential misuse that happen in the future.

Reply:
We can add new api's when increasing the minor version so I think adding a constructor overload which takes a `IDataProtectionProvider` is probably the cleanest way to do this. Then in the code you can call CreateProtector(purpose). When we go to 2.0, we can remove the other constructors. I'm then not sure what to do about the other constructors though. I can think of two options.   

1. Mark them as obsolete and throw an exception saying to use the new constructor passing an IDataProtectionProvider
2. On non-windows, throw PlatformNotSupportedException (which is what would happen if you tried to use the class on Linux today), but on Windows, if the existing constructor is used, still use ProtectedData.

The problem with option 1 is we intend to maintain backwards compatibility for compilation and if someone is referencing the older constructor they will get a compiler warning due to the obsolete attribute, and if they have warn as error turned on, it's breaking that promise. I doubt anyone today is creating an instance, but it is still technically going against the policy. I want to avoid making a precedent so early on. The problem with option 2 is that we keep a dependency on the ProtectedData package for a theoretical scenario which is almost certain to not ever be an issue.  

Having typed all this out, my current preference is a modified version of 1, but without the obsolete attribute. It maintains compilation compatibility, I'm 99.9999999% certain it won't break anyone, and if it does, we can go option 2 after to fix them. I'm not wanting to be a dictator here, so feel free to express your thoughts.